### PR TITLE
Add methods to detect Citus extension in a new Citus module

### PR DESCRIPTION
**Pull request by naisila**
A new module called "Citus" is added to the base methods, and it contains the methods for detecting citus extension.

### CODE REVIEW

File: app/views/layouts/pg_hero/application.html.erb [hunk 1]
```diff
@@ -18,7 +18,11 @@
         <% elsif Rails.env.development? %>
           Do not use development information to make decisions about your production environment
         <% else %>
-          <%= link_to "PgHero", root_path %>
+          <% if @citus_enabled %>
```
Comment (byucesoy) [inline app/views/layouts/pg_hero/application.html.erb]: I think we can remove this "feature" :)

File: lib/pghero/methods/citus.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      def citus_available?
+        select_one("SELECT COUNT(*) AS count FROM pg_available_extensions WHERE name = 'citus'") > 0
+      end
+
+      # only cache if true
+      def citus_enabled?
+        @citus_enabled ||= citus_readable?
+      end
+
+      def citus_extension_enabled?
```
Comment (byucesoy) [inline lib/pghero/methods/citus.rb]: Who uses this function?

File: lib/pghero/methods/citus.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module PgHero
+  module Methods
+    module Citus
+      def citus_version
+        @citus_version ||= select_one("SHOW citus.version")
+      end
+
+      def citus_available?
```
Comment (byucesoy) [inline lib/pghero/methods/citus.rb]: Who uses this function?

File: lib/pghero/methods/citus.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        select_one("SELECT COUNT(*) AS count FROM pg_available_extensions WHERE name = 'citus'") > 0
+      end
+
+      # only cache if true
```
Comment (byucesoy) [inline lib/pghero/methods/citus.rb]: It would be better if you write full sentences for comments.

File: app/controllers/pg_hero/home_controller.rb [hunk 5]
```diff
@@ -23,7 +23,8 @@ class HomeController < ActionController::Base
     def index
       @title = "Overview"
       @extended = params[:extended]
-
+      @citus_enabled = @database.citus_enabled?
```
Comment (byucesoy) [inline app/controllers/pg_hero/home_controller.rb]: I think we should add those lines with the PRs of each customized page, not with this PR.

File: app/controllers/pg_hero/home_controller.rb [hunk 6]
```diff
@@ -259,21 +265,37 @@ def tune
...
     def connections
       @title = "Connections"
       @connection_sources = @database.connection_sources
       @total_connections = @connection_sources.sum { |cs| cs[:total_connections] }
-
+      @citus_enabled = @database.citus_enabled?
+      
       @connections_by_database = group_connections(@connection_sources, :database)
       @connections_by_user = group_connections(@connection_sources, :user)
+      if @citus_enabled
```
Comment (byucesoy) [inline app/controllers/pg_hero/home_controller.rb]: I believe these changes belongs to another PR.

**Review by byucesoy — CHANGES_REQUESTED** [review]
Could you also change commit title so that;
* It starts with uppercase.
* It starts with a verb.

Some examples; https://github.com/citusdata/citus/commits/master

File: lib/pghero/methods/citus.rb (hunk 2 above)
Comment (naisila) [reply lib/pghero/methods/citus.rb]: It is not used for now so I will remove it.

File: lib/pghero/methods/citus.rb (hunk 3 above)
Comment (naisila) [reply lib/pghero/methods/citus.rb]: It is not used for now so I will remove it.

File: lib/pghero/methods/citus.rb (hunk 4 above)
Comment (naisila) [reply lib/pghero/methods/citus.rb]: Ok.

File: app/views/layouts/pg_hero/application.html.erb (hunk 1 above)
Comment (naisila) [reply app/views/layouts/pg_hero/application.html.erb]: Ok.

File: app/controllers/pg_hero/home_controller.rb (hunk 5 above)
Comment (naisila) [reply app/controllers/pg_hero/home_controller.rb]: Ok.

File: app/controllers/pg_hero/home_controller.rb (hunk 6 above)
Comment (naisila) [reply app/controllers/pg_hero/home_controller.rb]: Yes they do. I accidentally put them here.

**Comment by naisila** [comment]
@byucesoy I have implemented the changes you suggested.

File: lib/pghero/methods/citus.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+module PgHero
+  module Methods
+    module Citus
+      # It will only cache citus_enabled if it is true once.
+      def citus_enabled?
+        @citus_enabled ||= citus_readable?
+      end
+
+      def citus_readable?
+        select_all("SELECT 1 FROM pg_dist_partition LIMIT 1")
```
Comment (byucesoy) [inline lib/pghero/methods/citus.rb]: I think it would be better if we use something like;

```sql
SELECT EXISTS(SELECT * FROM pg_available_extensions() WHERE name = 'citus');
```

That would probably save us from except'on handling too.

**Review by byucesoy — CHANGES_REQUESTED** [review]
I think this PR looks OK. It only needs tests.

File: lib/pghero/methods/citus.rb (hunk 7 above)
Comment (naisila) [reply lib/pghero/methods/citus.rb]: `pg_extension` would be appropriate because `citus` may be available but not installed yet. I have used this way to make sure `pg_dist_partition` is there. I wonder, is there any possibility that `citus` extension has been created but somehow `pg_dist_partition` is not there?

Comment (byucesoy) [reply lib/pghero/methods/citus.rb]: Right, `pg_extension` is a better choice. I think we should not rely on the existence of `pg_dist_partition`, because;

- Although it is rare, we can change catalog tables in the future.
- We have other catalog tables as well, randomly picking one feels awkward
- Using exception handling for checking availability is bad pattern.

Comment (naisila) [reply lib/pghero/methods/citus.rb]: I agree on the third point. Regarding the first two, I am not randomly picking one. Maybe I should add a comment that explains that. `pg_dist_partition` is a crucial catalog to make PgHero Citus-compatible, and almost every method uses it. Changing this catalog table would require changing the methods as well anyway. If somehow `pg_dist_partition` is not there, the respective methods would error out. Rather than checking for its existence in all the methods, I have chosen to check it in `citus_enabled`. However, if it is extremely rare that `citus` extension is created and `pg_dist_partition` not, then we should use `pg_extension` because, as you said, it saves from exception handling.

**Comment by naisila** [comment]
@byucesoy I have added the tests.

File: Rakefile [hunk 8]
```diff
@@ -6,3 +6,9 @@ Rake::TestTask.new do |t|
   t.libs << "test"
   t.pattern = "test/**/*_test.rb"
 end
+
+task :test2
```
Comment (byucesoy) [inline Rakefile]: Can we give this more meaningful name?

File: Rakefile [hunk 9]
```diff
@@ -6,3 +6,9 @@ Rake::TestTask.new do |t|
...
+Rake::TestTask.new do |t|
+  t.libs << "test"
+  t.pattern = "test/**/*_test_citus.rb"
+end
```
Comment (byucesoy) [inline Rakefile]: ~~In unix based systems, files should have one blank line at the end. This weird symbol at the end indicates that there is no blank line.~~

just ignore the above comment.

File: Rakefile (hunk 8 above)
Comment (byucesoy) [reply Rakefile]: Also when I execute the command `bundle exec rake test2` nothing happens. Could you please check that.

**Review by byucesoy — CHANGES_REQUESTED** [review]
Could you please ensure that executing `bundle exec rake test2` actually runs citus tests.

File: test/basic_test_citus.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require_relative "test_citus_helper"
+
+class BasicCitusTest < Minitest::Test
+  def test_citus
+    assert_equal true, PgHero.citus_enabled?
+  end
+end
```
Comment (byucesoy) [inline test/basic_test_citus.rb]: This needs blank line too.

File: Rakefile (hunk 8 above)
Comment (naisila) [reply Rakefile]: Done 👍

**Review by byucesoy — APPROVED** [review]
Looks great. :shipit:
